### PR TITLE
feat: implements init_dataset command

### LINKED ISSUE #12: Implement init_database command
Initialize a dataset

### PR DESCRIPTION
**Pull request by Auroric**
This PR implements the init_dataset command. 

Closes #12

### CODE REVIEW

**Comment by Auroric** [comment]
Making this draft PR to make sure I'm going in the right direction w.r.t. what @falafel72 mentioned last meeting. It compiles, but nothing has been tested so far. My main concerns as of right now are:
1. What is the default directory if none is provided when initializing a dataset? I've set it as the home folder for now, but the `e4edm` [documentation](https://github.com/UCSD-E4E/e4e-data-management/blob/cb5168d1823302bbd2d5181fdc044a3980b49e9e/README.md?plain=1#L23) mentions the user's data directory, set in the config. Does that correspond to the config's `dataset_dir`? So far it seems that it's an optional type, so what happens if it's not set?
2. Some commands necessarily need to modify the config, but the struct members are all private and the config itself is not mutable. I've changed the necessary ones to public and specifically marked the config mutable to make this command work. Should the config be modified in a different way than this?
3. So far it looks like there's no way to initialize `Dataset` (and also `Manifest`) objects. This PR implements very basic builders for them, but is there a better way to do this?

**Comment by ntlhui** [comment]
When `e4edm` initializes a dataset, we by default select `user_data_dir` (see https://github.com/UCSD-E4E/e4e-data-management/blob/cb5168d1823302bbd2d5181fdc044a3980b49e9e/e4e_data_management/core.py#L37).  This is from [appdirs](https://pypi.org/project/appdirs/).  The `rust` equivalent at this point appears to be [directories](https://crates.io/crates/directories).

Functionally, this should be a location (per OS standards) that is specific to this program, persistent, and nominally without a quota.

This must be set - initializing a dataset should immediately create the appropriate directory structure and empty dataset datastructures.

**Comment by falafel72** [comment]
2. Making the reference to the config mutable is probably how it should be. The app should maintain a single instance of the config struct which is modified using the reference. 
3. Both of these should either have a default initialization (i.e. the empty case) or should be loaded in from files, so a basic builder works just fine. `init_dataset` doesn't implement the file loading case, that will likely be one of the next issues I add.

**Comment by Auroric** [comment]
> When `e4edm` initializes a dataset, we by default select `user_data_dir` (see https://github.com/UCSD-E4E/e4e-data-management/blob/cb5168d1823302bbd2d5181fdc044a3980b49e9e/e4e_data_management/core.py#L37). This is from [appdirs](https://pypi.org/project/appdirs/). The `rust` equivalent at this point appears to be [directories](https://crates.io/crates/directories).
> 
> Functionally, this should be a location (per OS standards) that is specific to this program, persistent, and nominally without a quota.
> 
> This must be set - initializing a dataset should immediately create the appropriate directory structure and empty dataset datastructures.

@ntlhui I read through the code and the AppDirs library [documentation](https://pypi.org/project/appdirs/) but it isn't particularly clear on how it converts the app author and app name referenced [here](https://github.com/UCSD-E4E/e4e-data-management/blob/cb5168d1823302bbd2d5181fdc044a3980b49e9e/e4e_data_management/core.py#L26) to a directory structure, particularly regarding whether it preserves spaces in the directory path. I presume we would want the Rust rewrite of `e4edm` to be compatible with the config structure of the Python version, so would you be able to provide the appropriate path generated by Python `e4edm` to emulate?

**Comment by ntlhui** [comment]
No, we don't need to maintain backwards compatibility. The previous uses pickle to serialize state data anyways, and that can't be ported to rust

File: src/cli.rs [hunk 1]
```diff
@@ -1,6 +1,6 @@
-use crate::commands::Commands;
+use crate::commands::{Commands, init_dataset::init_dataset};
 use crate::config::E4EDMConfig;
-use anyhow::Result;
+use anyhow::{Result, bail};
```
Comment (falafel72) [inline src/cli.rs:3]: bail is really nice, thanks for adding it

File: src/commands/init_dataset.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    );
+    let dataset_path = args.path.to_owned()
+        .unwrap_or(ProjectDirs::from("com", "Engineers For Exploration", "E4EDataManagement").unwrap().config_dir().to_path_buf())
+        .join(dataset_name.clone());
+
+    if config.datasets.contains_key(&dataset_name) {
+        bail!("Dataset with that name already exists!");
+    }
+    config.active_dataset = Some(dataset_name.clone());
+    config.datasets.insert(dataset_name, build_dataset(dataset_path, args.date.with_timezone(&Utc)));
```
Comment (falafel72) [inline src/commands/init_dataset.rs:24]: Current behavior of the python app is to save the dataset after it is built (so make the folder and add a blank manifest). We'll need to add a `save` function to the `Dataset` struct, but we can implement what it does later on. 

Also, I don't know if it makes sense to store the UTC timestamp. My thought we can probably just get away with using the date (the `NaiveDate` type in `chrono`) since this the dataset is meant to have more coarse-grain information anyway.

File: src/cli.rs [hunk 3]
```diff
@@ -1,6 +1,6 @@
-use crate::commands::Commands;
+use crate::commands::{Commands, init_dataset::init_dataset};
```
Comment (falafel72) [inline src/cli.rs:1]: This module structure is not great, which is my doing, not yours - the way we structure things right now forces us to do it this way. I'm going to say this is fine for now and probably do some refactoring on my own.

File: src/dataset.rs [hunk 4]
```diff
@@ -25,6 +26,27 @@ pub struct Dataset {
     version: String,
 }
 
+pub fn build_dataset(root: std::path::PathBuf, day_0: DateTime<Utc>) -> Dataset {
```
Comment (falafel72) [inline src/dataset.rs:29]: See comments in `init_dataset.rs`

**Review by falafel72 — CHANGES_REQUESTED** [review]
tl;dr my recommendation is to change the `day_0` field in the `Dataset` struct to a `NaiveDate`, see comments

File: src/commands/init_dataset.rs (hunk 2 above)
Comment (Auroric) [reply src/commands/init_dataset.rs:24]: > Also, I don't know if it makes sense to store the UTC timestamp. My thought we can probably just get away with using the date (the `NaiveDate` type in `chrono`) since this the dataset is meant to have more coarse-grain information anyway.

@falafel72 I've made the changes in my most recent commit ([7b903e4](https://github.com/UCSD-E4E/e4e-data-management-rust/pull/13/commits/7b903e4510a8c1c2fac31e2b5661fab386975caf)) but wasn't sure if this changed something with `serde` since I removed one of the annotations while making this change, and I'm not familiar enough with it to know exactly what that line did. I'm aware @cgcha did some stuff with `serde` -- could you verify that what I've done is ok?

**Comment by ntlhui** [comment]
I'm not sure that you can make the statement that you should use a naive timestamp due to coarse grained formatting. Timezones creates an offset, so it is possible that you format the wrong date. @ccrutchf can you comment further?

**Comment by ccrutchf** [comment]
I agree.  You need to maintain timezones basically always since everything will be relative to that date.  Otherwise, you lose context if your dataset is in a different timezone or if you move timezones during collecting data, i.e. daylight savings.

**Comment by falafel72** [comment]
Very well, the codebase currently won't compile if we allow for any timezone since we haven't defined how to serialize them yet, so it's blocked on #6. Changing it back to UTC will have to do for now, follow this stack overflow post for how to undo your last commit (sorry): https://stackoverflow.com/questions/36168839/how-to-remove-commits-from-a-pull-request

**Comment by ntlhui** [comment]
I would recommend that we store these as ISO 8601 or RFC 3339 formatted
strings.

Note that JSON Schema has explicit support for RFC3339.

Nathan T. Hui, MSc
R&D Engineer
UCSD E4E/CHEI
tel: 858.534.9306 | ***@***.***

**Review by falafel72 — APPROVED** [review]
